--- ./amarok-1.4.10/amarok/src/metadata/mp4/mp4properties.cpp	2010-06-04 01:43:50.859898428 -0700
+++ ./amarok-1.4.10/amarok/src/metadata/mp4/mp4properties.cpp	2010-06-04 01:42:51.360657950 -0700
@@ -23,6 +23,7 @@ email                : andrew.leadbetter
 
 
 #include <tstring.h>
+#include <cstring>
 
 #include <config.h>
 #ifdef HAVE_SYSTEMS_H
--- ./amarok-1.4.10/amarok/src/metadata/mp4/taglib_mp4filetyperesolver.cpp	2010-06-04 01:43:43.456659525 -0700
+++ ./amarok-1.4.10/amarok/src/metadata/mp4/taglib_mp4filetyperesolver.cpp	2010-06-04 01:42:29.056923522 -0700
@@ -24,6 +24,7 @@
 
 #include "taglib_mp4filetyperesolver.h"
 #include "mp4file.h"
+#include <cstring>
 
 TagLib::File *MP4FileTypeResolver::createFile(const char *fileName,
         bool readProperties,
